client: Support concurrent batch execution on client.Txn

nvb opened the pull request:
Fixes #13684.
First step of #13160.

Up to this point, `client.Txn` has required that only a single `client.Batch` be
executed at a time. It had no support for concurrent requests, and was not safe
to use by multiple goroutines. This change lifts this restriction by adding
necessary synchronization into `client.Txn` methods and restructuring request
transmission logic.

One of the biggest changes is that Transaction initialization has been moved to
the client package. This is necessary because the initialization has to be done
under lock. Initialization was also tricky because we needed to avoid sending
multiple `BeginTransactionRequests`. Ben's [comment](https://github.com/cockroachdb/cockroach/issues/13684#issuecomment-281427763) suggests that it is sufficient
to have only a single request contain the BeginTransaction as long as it is at
least concurrent with the first write. To make sure of this, we now maintain a
`writingTxnRecord` flag on `client.Txn` that works in conjunction with the
`Proto.Writing` flag to determine when a BeginTransaction request needs to
be inserted into a batch.

One difficulty was that concurrent requests greatly increase the chance that
we'll reset the transaction Proto on an abort before a later error is received
for that transaction ID. This meant that our `IsRetryableErrMeantForTxn`
handling became much more important, and that the previous TODOs about ID
detection after Proto reset needed to be addressed. To combat this, we now
maintain a map of past IDs on the `client.Txn` object. This allows us to check
if an error is addressed to any incarnation of the transaction.

Finally, this change and its complications with transaction initialization
necessitated a few changes to `TxnCoordSender`. There were a few assertions
about the `Writing` state of a `roachpb.Transaction` that needed to be removed
to handle situations where concurrent requests are issued before a transaction
has written its transaction record. This `Writing` state was previously used to
enforce that only one coordinator can be used for transactional writes, so for
now I've commented out `TestTxnMultipleCoord`. I'd appreciate input on this
issue.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13799)
<!-- Reviewable:end -->

nvb committed:
client: Give client.DB its very own clock

This clock will be used to assign timestamps to client.Txns once that
responsibility moves to the client package.

nvb committed:
client: Fix TestTransactionKeyNotChangedInRestart

The test was previously asserting the exact opposite of what it claimed
to test; it was testing that the transaction anchor keys changed between
restarts. This change fixes the test, making it actually assert
that the transaction key does not get changed during a restart.

nvb committed:
client: Support concurrent batch execution on client.Txn

Fixes #13684.
First step for #13160.

Up to this point, `client.Txn` has required that only a single `client.Batch`
be executed at a time. It had no support for concurrent requests, and was
not safe to use by multiple goroutines. This change lifts this restriction
by adding necessary synchronization into `client.Txn` methods and restructuring
request transmission logic.

One of the biggest changes is that Transaction initialization has been
moved to the client package. This is necessary because the initialization
has to be done under lock. Initialization was also tricky because we needed
to avoid sending multiple `BeginTransactionRequests`. Ben's
[comment](https://github.com/cockroachdb/cockroach/issues/13684#issuecomment-281427763)
suggested that it would be sufficient to have only a single request contain the
BeginTransaction as long as it is at least concurrent with the first write.
To assert this, we now maintain a `writingTxnRecord` flag on `client.Txn`
that works in conjunction with the `Proto.Writing` flag to determine when
a BeginTransaction request needs to be created.

One difficulty was that concurrent requests greatly increase the chance
that we'll reset the transaction Proto on an abort before a later error
is received for that transaction ID. This meant that our `IsRetryableErrMeantForTxn`
handling became much more important, and that the previous TODOs about ID detection
after Proto reset needed to be addressed. To combat this, we now maintain a map
of past IDs on the `client.Txn` object. This allows us to check if an error
is addressed to any incarnation of the transaction.

Finally, this change and its complications with transaction initialization
necessitated a few changes to `TxnCoordSender`. There were a few assertions
about the `Writing` state of a `roachpb.Transaction` that needed to be removed
to handle situations where concurrent requests are issued before a transaction
has written its transaction record. This `Writing` state was previously used
to enforce that only one coordinator can be used for transactional writes, so
for now I've commented out `TestTxnMultipleCoord`. I'd appreciate input on
this issue.

bdarnell commented:
Reviewed 19 of 23 files at r2, 19 of 19 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxcNt9NViPkyDLHoU3:-KdxcNtAv7sBerSK2kcF:b-t9u4gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L116)):*
> ```Go
> 
> func (txn *Txn) isFinalizedLocked() bool {
> 	return txn.mu.finalized
> ```

I think for a trivial (and unexported) accessor like this I'd just inline the direct access to `mu.finalized`.

---

*[pkg/internal/client/txn.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxeIhBfw5ux4zAz-ZH:-KdxeIhBfw5ux4zAz-ZI:bk9fgbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L173)):*
> ```Go
> //
> 
> // Proto not thread-safe, only use if you know that no requests are executing concurrently.
> ```

These should probably have names that indicate their thread-unsafety. Or can we just make them safe by grabbing the lock? That wouldn't work for all of them (e.g. `Proto` returns a pointer to memory under the lock), but `DebugName` and `Isolation` look like they could get the lock.

---

*[pkg/internal/client/txn.go, line 607 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Kdxh3CFnU-cpgGOlUWs:-Kdxh3CFnU-cpgGOlUWt:b-5a2xda) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L607)):*
> ```Go
> 				// EndTxnRequest.
> 				//
> 				// REVIEW: Is that an issue?
> ```

What other EndTxnRequest could this race with? I think we probably need to guarantee that `Exec` with `AutoCommit` is only used in situations where nothing else could try to commit the transaction. 

---

*[pkg/kv/txn_coord_sender.go, line 523 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjIXSxFPHzpiR8TQv:-KdxjIXTQiOxl6LxDKLG:bvefpfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender.go#L523)):*
> ```Go
> 	}
> 	if !ba.Txn.IsInitialized() {
> 		// This is not possible if the BatchRequest is sent from a client.Txn,
> ```

So is this just for tests now?

---

*[pkg/kv/txn_coord_sender.go, line 545 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjOzJoOi5BLiRDkEJ:-KdxjOzJoOi5BLiRDkEK:b-1xg79s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender.go#L545)):*
> ```Go
> 			}
> 		}
> 		// FOR REVIEW: we can't have this check anymore, because concurrent
> ```

I'm OK with removing it.

---

*[pkg/kv/txn_coord_sender_test.go, line 956 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjvEpIIxkObldkkOx:-KdxjvEquQ0OsB6mAjOs:b-kkida4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender_test.go#L956)):*
> ```Go
> //
> //   I think we may have had to change this to support running transactions from
> //   multiple nodes anyway, so it might be a good time to rethink this...
> ```

I think it's fine to make `maybeRejectClientLocked` a little less sensitive: until we get back the txn with `Writing` set, it's OK if we can't requests from transactions that have expired or attempted to use multiple coordinators. But we should still make the best-effort check once `Writing` is set, so I think it should be possible to revive this test. 

---

*[pkg/roachpb/data.go, line 662 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxlMliK34-ietqA9EI:-KdxlMliK34-ietqA9EJ:b-q9cjot) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/roachpb/data.go#L662)):*
> ```Go
> )
> 
> // NewTransaction behaves like MakeTransaction, but returns the new
> ```

We rarely have both `New` and `Make` variations for the same type. I think it's better to just pick one and inline the reference/dereference logic for the uncommon case. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 416 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-Q697N_44HP6rgJDg:-Ke-Q6982lveg-eQ5Yob:b-apvh3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L416)):*
> ```Go
> 		panic("no error")
> 	}
> 	// This may race with a concurrent EndTxnRequests. That's fine though because
> ```

can you comment that the other `EndTxnRequest`s are assumed to also be rollbacks?

Do we actually need to allow this method to be called concurrently? Shouldn't the client (i.e. sql) synchronize these state transitions (committing and rolling back)?

If we do need to allow this call concurrently,  I think I would feel better if we improved the code to no actually send concurrent `EndTxnRequest`s (e.g. by having a channel to wait on for the results of pending `EndTxnRequest`s). And we should also document and enforce whatever assumptions we make about the client calling `Commit` and `Rollback` concurrently.
What do you think?

---

*[pkg/internal/client/txn.go, line 538 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-SEZ2h3Fvxj-m2l3v:-Ke-SEZ2h3Fvxj-m2l3w:b-7fxc1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L538)):*
> ```Go
> 	// ordering.
> 	//
> 	// REVIEW: is there a better name for this?
> ```

note to not forget about this

I'm not sure yet about a better name (I suspect that the name should suggest the choice about _who_ assigns the `OrigTimestamp`), but I think this field definitely deserves a better comment. Who may want to set it and who may want not to (and if not set, who assigns the `OrigTimestamp`)?
Haven't checked the code around this field but I'm surrpised that, given that the txns start being used concurrently, we don't have to always set the `OrigTimestamp` before allowing the txn to send any requests...

---

*[pkg/internal/client/txn.go, line 607 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Kdxh3CFnU-cpgGOlUWs:-Ke-TxKSG3wvEr-79hKU:b-mpr1lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L607)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What other EndTxnRequest could this race with? I think we probably need to guarantee that `Exec` with `AutoCommit` is only used in situations where nothing else could try to commit the transaction. 
</blockquote></details>

+1. I had similar comment on the `Commit/Rollback` methods.

---

*[pkg/internal/client/txn_test.go, line 473 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-NRAIP7erJKP-5Z3x:-Ke-NRAJzzs5GrppMW0v:b-snnrfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L473)):*
> ```Go
> 
> // TestTransactionKeyNotChangedInRestart verifies that if the transaction already has a key (we're
> // in a restart), the key in the begin transaction request is not changed.
> ```

maybe expand this comment to explain that changing the (first) key the txn writes to between retries is the key

---

*[pkg/internal/client/txn_test.go, line 512 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-MqDSCYrP_Z7BOQyD:-Ke-MqDTKW6DU4bGF0uP:b-we0wzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L512)):*
> ```Go
> 		}
> 		return ba.CreateReply(), nil
> 	}, nil))
> ```

please comment the `nil` with /* post */ 

---

*[pkg/internal/client/txn_test.go, line 517 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-N9hnBVKWBDFtgTaT:-Ke-N9hnBVKWBDFtgTaU:bar3aed) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L517)):*
> ```Go
> 		defer func() { attempt++ }()
> 		b := txn.NewBatch()
> 		b.Put(keys[0], "b")
> ```

I think you want `keys[attempt]`, right? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

nvb committed:
kv: Completely remove transaction initialization from TxnCoordSender

Previously we had logic in `TxnCoordSender` to initialize transaction on
`BatchRequests` if they were not already initialized. This logic has
been removed and the few cases where tests depended on it have been
updated.

nvb committed:
client: Revive TestTxnMultipleCoord

nvanbenschoten commented:
Review status: 0 of 38 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/internal/client/txn.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxcNt9NViPkyDLHoU3:-Ke1pZzBQ3Y-jal5bBdU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L116)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think for a trivial (and unexported) accessor like this I'd just inline the direct access to `mu.finalized`.
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxeIhBfw5ux4zAz-ZH:-Ke28LnfyZvKTo_lbAX6:brxy7sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L173)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These should probably have names that indicate their thread-unsafety. Or can we just make them safe by grabbing the lock? That wouldn't work for all of them (e.g. `Proto` returns a pointer to memory under the lock), but `DebugName` and `Isolation` look like they could get the lock.
</blockquote></details>

I made most of these methods thread safe so that we dont need the distinction. 

---

*[pkg/internal/client/txn.go, line 416 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-Q697N_44HP6rgJDg:-Ke2HkXG2XOR8SFI2Ku0:bjkkapv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L416)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can you comment that the other `EndTxnRequest`s are assumed to also be rollbacks?

Do we actually need to allow this method to be called concurrently? Shouldn't the client (i.e. sql) synchronize these state transitions (committing and rolling back)?

If we do need to allow this call concurrently,  I think I would feel better if we improved the code to no actually send concurrent `EndTxnRequest`s (e.g. by having a channel to wait on for the results of pending `EndTxnRequest`s). And we should also document and enforce whatever assumptions we make about the client calling `Commit` and `Rollback` concurrently.
What do you think?
</blockquote></details>

It is possible that the `TxnCoordSender` may send an `EndTransactionRequest` in certain situations (`tryAsyncAbort`). I don't think we'll be able to guarantee at this level alone that only one `EndTransactionRequest` is ever in flight at a time.

With respect to the SQL layer, yes you're correct that it will always synchronize and send a single state transition. Would documenting: `"the result of sending concurrent commit and rollback requests is undefined"` be good enough?

---

*[pkg/internal/client/txn.go, line 538 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-SEZ2h3Fvxj-m2l3v:-Ke2JXsnmMRl86hhbtNm:b-qggohx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L538)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

note to not forget about this

I'm not sure yet about a better name (I suspect that the name should suggest the choice about _who_ assigns the `OrigTimestamp`), but I think this field definitely deserves a better comment. Who may want to set it and who may want not to (and if not set, who assigns the `OrigTimestamp`)?
Haven't checked the code around this field but I'm surrpised that, given that the txns start being used concurrently, we don't have to always set the `OrigTimestamp` before allowing the txn to send any requests...
</blockquote></details>

Your point at the end really scared me for a second... We do actually set the `OrigTimestamp` when we call `NewTransaction`, so the field will be set before any concurrent requests are sent out.

You were the last one to [touch the field](https://github.com/cockroachdb/cockroach/commit/aed470ef70557c39324f93514009616a64d90a06), so I'll assume you're the expert on it :trollface:. From what I can tell, the purpose of the field is solely to decide on an `OrigTimestamp` before `Txn.Exec` calls its provided closure (as opposed to when the first KV request is sent), so that the timestamp can serve as the lower bound for all actions performed in that closure. If that's true, maybe we name it `AssignTimestampImmediately`.

---

*[pkg/internal/client/txn.go, line 607 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Kdxh3CFnU-cpgGOlUWs:-Ke2NfVn4p3iVNk1A10s:bqilioi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L607)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

+1. I had similar comment on the `Commit/Rollback` methods.
</blockquote></details>

Couldn't the `TxnCoordSender` try to abort a transaction while we're here? I'm not super clear on the different ways that a transaction can be completed, so I'd appreciate any insight.

Also, by guarantee do you just mean include a comment or is there some way you envision enforcing that?

---

*[pkg/internal/client/txn_test.go, line 473 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-NRAIP7erJKP-5Z3x:-Ke1uFN4DksUStafPyeJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L473)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe expand this comment to explain that changing the (first) key the txn writes to between retries is the key
</blockquote></details>

Done.

---

*[pkg/internal/client/txn_test.go, line 512 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-MqDSCYrP_Z7BOQyD:-Ke1uMDQ54VU1fQyEuLb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L512)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please comment the `nil` with /* post */ 
</blockquote></details>

Done.

---

*[pkg/internal/client/txn_test.go, line 517 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-N9hnBVKWBDFtgTaT:-Ke1tj_6vVhAD6Us9sph:bt8447p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L517)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think you want `keys[attempt]`, right? 
</blockquote></details>

Wow, yeah. This test just does not want to be correct. I'm even more confused now, though. We only restart a transaction completely from scratch (send a new `BeginTxnRequest`) in the case of a `TransactionAbortedError`. However, our `TransactionAbortedError` [never retained](https://github.com/cockroachdb/cockroach/blob/1222d5d553b2fee93b572c6513d4fddba3c45d89/pkg/internal/client/txn.go#L621) the transaction key across aborts. Regardless of the broken test, I'm confused how this ever worked. Are we supposed to resend `BeginTxnRequests` with the same key in cases where they are retransmitted due to an abort (in which case we've never done this, unless I'm completely missing something), or was #5492 misinterpreted to begin with?

---

*[pkg/kv/txn_coord_sender.go, line 523 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjIXSxFPHzpiR8TQv:-Ke2d8oC2FWt8QOEBGlM:b6z7aok) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender.go#L523)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

So is this just for tests now?
</blockquote></details>

Yes. I've added another commit that fully removes transaction initialization from the `TxnCoordSender`.

---

*[pkg/kv/txn_coord_sender.go, line 545 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjOzJoOi5BLiRDkEJ:-Ke23TGUjRDEjiVO3ZKi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender.go#L545)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm OK with removing it.
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender_test.go, line 956 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjvEpIIxkObldkkOx:-Ke2g1DcdVUoof2h4sOw:btyonpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender_test.go#L956)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think it's fine to make `maybeRejectClientLocked` a little less sensitive: until we get back the txn with `Writing` set, it's OK if we can't requests from transactions that have expired or attempted to use multiple coordinators. But we should still make the best-effort check once `Writing` is set, so I think it should be possible to revive this test. 
</blockquote></details>

Done. Not 100% confident that the test is asserting the behavior we want, though.

---

*[pkg/roachpb/data.go, line 662 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxlMliK34-ietqA9EI:-Ke23vQJoN9qSbV3Z8hl:bvcu55) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/roachpb/data.go#L662)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We rarely have both `New` and `Make` variations for the same type. I think it's better to just pick one and inline the reference/dereference logic for the uncommon case. 
</blockquote></details>

I was hoping to avoid pointless allocations where possible. Probably better to avoid perf concerns while this larger change is going on, though. Reverted to just `NewTransaction`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
We discussed moving the txn proto initialization to the `client.Txn` constructor. Seems like it would simplify things - e.g. get rid of the `AssignOrigTimestamp`field.

---

Review status: 0 of 38 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 92 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke4d4pK0veFdy0FCArP:-Ke4d4pK0veFdy0FCArQ:b7gcqor) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9c4e1bac0d2e2f62fa152db39d1ed116fa373be/pkg/internal/client/txn.go#L92)):*
> ```Go
> }
> 
> // IsInitialized returns true if the transaction has been initialized.
> ```

maybe hang some more words from this comment about who initializes the proto.

---

*[pkg/kv/txn_coord_sender.go, line 523 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KdxjIXSxFPHzpiR8TQv:-Ke4eCm7_apI35ezOGoA:b-q74txj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/kv/txn_coord_sender.go#L523)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes. I've added another commit that fully removes transaction initialization from the `TxnCoordSender`.
</blockquote></details>

great!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-:-Ke4rpC2AZNTAq8-98iq:b7bot4k)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 19 of 23 files at r5, 15 of 18 files at r6, 2 of 4 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 416 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-Q697N_44HP6rgJDg:-Ke4odwl19wytmfhy315:b-98q6f7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L416)):*
> Would documenting: "the result of sending concurrent commit and rollback requests is undefined" be good enough?

The word "undefined" is scary (at least for anyone who's fought with a C compiler), but I don't think we need to use such strong language. I'm pretty sure that multiple *abort* attempts are fine; we just need to be careful with commits and regular commands. (and if a commit races with an abort, one or the other will win; you won't get some undefined mixture of the two)

So for a new warning, I'd say "It is undefined to call `Commit` concurrently with any call to `Exec`. Since `Exec` with the `AutoCommit`flag is equivalent to an `Exec` possibly followed by a `Commit`, it must not be called concurrently with any other call to `Exec` or `Commit`. "

---

*[pkg/internal/client/txn.go, line 538 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-SEZ2h3Fvxj-m2l3v:-Ke4t9P_5bLdc1V3m7Ju:bl2yo2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L538)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Your point at the end really scared me for a second... We do actually set the `OrigTimestamp` when we call `NewTransaction`, so the field will be set before any concurrent requests are sent out.

You were the last one to [touch the field](https://github.com/cockroachdb/cockroach/commit/aed470ef70557c39324f93514009616a64d90a06), so I'll assume you're the expert on it :trollface:. From what I can tell, the purpose of the field is solely to decide on an `OrigTimestamp` before `Txn.Exec` calls its provided closure (as opposed to when the first KV request is sent), so that the timestamp can serve as the lower bound for all actions performed in that closure. If that's true, maybe we name it `AssignTimestampImmediately`.
</blockquote></details>

@andreimatei and I talked about this offline. We think that this code dates back to when `client.Txn` was going to be a public interface and ported to many languages, so we wanted to keep as much logic as possible on the server (`TxnCoordSender`) side (and so we didn't need clocks for all the tests that use `client.Txn`). But now that `client.Txn` is internal-only and we require the presence of a clock, we might as well assign the orig timestamp unconditionally here (assuming that doesn't break too many tests). 

---

*[pkg/internal/client/txn.go, line 607 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Kdxh3CFnU-cpgGOlUWs:-Ke4uGAl6jjSpKT8EnKg:b5t4j4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L607)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Couldn't the `TxnCoordSender` try to abort a transaction while we're here? I'm not super clear on the different ways that a transaction can be completed, so I'd appreciate any insight.

Also, by guarantee do you just mean include a comment or is there some way you envision enforcing that?
</blockquote></details>

I think this "may race with another EndTxnRequest" is scarier than it should be (the other comments like this are fine because they're about aborts, but this one is a commit). I think I'd just remove the new sentence from this comment. 

To enforce this, we'd need to do something under the lock where we access `txn.Proto().Status` - set a flag indicating that we intend to finalize the transaction (here and in `CleanupOnError`. `Commit` would then need to check this flag too). There'd still be a chance that the lower-level `TxnCoordSender` could  still abort the transaction, but nothing else could commit it. This isn't strictly necessary but I think it might be a good idea to remove a little bit of potential raciness. 

---

*[pkg/internal/client/txn_test.go, line 517 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-N9hnBVKWBDFtgTaT:-Ke4vu1Es0jewKfv0YCY:b-17sfu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L517)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Wow, yeah. This test just does not want to be correct. I'm even more confused now, though. We only restart a transaction completely from scratch (send a new `BeginTxnRequest`) in the case of a `TransactionAbortedError`. However, our `TransactionAbortedError` [never retained](https://github.com/cockroachdb/cockroach/blob/1222d5d553b2fee93b572c6513d4fddba3c45d89/pkg/internal/client/txn.go#L621) the transaction key across aborts. Regardless of the broken test, I'm confused how this ever worked. Are we supposed to resend `BeginTxnRequests` with the same key in cases where they are retransmitted due to an abort (in which case we've never done this, unless I'm completely missing something), or was #5492 misinterpreted to begin with?
</blockquote></details>

`TransactionAbortedError` doesn't need to retain the key. `TransactionRetryError` *must* retain the key and *must not* send another `BeginTransaction` request. (That would require a pretty strange sequence of events; maybe it's a regression test for a bug we saw at some point?). So we want a transaction that writes to "first" on the first attempt (sending BeginTransaction, Put, and then getting a TransactionRetryError), then writes to "second" on the second iteration (sending Put with a Txn whose key points to "first") then commits. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 38 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/internal/client/txn.go, line 416 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-Q697N_44HP6rgJDg:-KeD3twLbPk4Pi5YD88N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L416)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> Would documenting: "the result of sending concurrent commit and rollback requests is undefined" be good enough?

The word "undefined" is scary (at least for anyone who's fought with a C compiler), but I don't think we need to use such strong language. I'm pretty sure that multiple *abort* attempts are fine; we just need to be careful with commits and regular commands. (and if a commit races with an abort, one or the other will win; you won't get some undefined mixture of the two)

So for a new warning, I'd say "It is undefined to call `Commit` concurrently with any call to `Exec`. Since `Exec` with the `AutoCommit`flag is equivalent to an `Exec` possibly followed by a `Commit`, it must not be called concurrently with any other call to `Exec` or `Commit`. "
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 538 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-SEZ2h3Fvxj-m2l3v:-KeQnYcWx-iUANATImsI:b-qannhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L538)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

@andreimatei and I talked about this offline. We think that this code dates back to when `client.Txn` was going to be a public interface and ported to many languages, so we wanted to keep as much logic as possible on the server (`TxnCoordSender`) side (and so we didn't need clocks for all the tests that use `client.Txn`). But now that `client.Txn` is internal-only and we require the presence of a clock, we might as well assign the orig timestamp unconditionally here (assuming that doesn't break too many tests). 
</blockquote></details>

I've opened up the issue #13894 and have a working fix at https://github.com/nvanbenschoten/cockroach/commit/363b9ddd1093cbda253623a91ffc52dff8930120. This will be a decently intrusive change because of some of the past assumptions it will break regarding transaction initialization, and I think we all agreed it should be separated from this PR. For now, I'm going to rename `AssignOrigTimestamp` to `AssignTimestampImmediately` (and add a better comment) until we can get rid of it. 

---

*[pkg/internal/client/txn.go, line 607 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Kdxh3CFnU-cpgGOlUWs:-KeQsTnbB6uuCcFGOWEt:b-kgowk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fadbbd9fb18ba4999472cf46365dc6db8443dc67/pkg/internal/client/txn.go#L607)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think this "may race with another EndTxnRequest" is scarier than it should be (the other comments like this are fine because they're about aborts, but this one is a commit). I think I'd just remove the new sentence from this comment. 

To enforce this, we'd need to do something under the lock where we access `txn.Proto().Status` - set a flag indicating that we intend to finalize the transaction (here and in `CleanupOnError`. `Commit` would then need to check this flag too). There'd still be a chance that the lower-level `TxnCoordSender` could  still abort the transaction, but nothing else could commit it. This isn't strictly necessary but I think it might be a good idea to remove a little bit of potential raciness. 
</blockquote></details>

@andreimatei and I spent some time discussing #10511 and what will need to be done to address that issue. Merging transaction heartbeats into `client.Txn` was one of the biggest issues, but once we address that through some extra layer of locking we can concretely assure that only one `EndTxnRequest` is ever sent at a time. Since performing this kind of locking right now could not provide this guarantee, and since the raciness should never be seen given our current usage patterns of `Txn.Exec` (we never operate on the `Txn` object concurrently outside of the method's closure, and just added the warning not to `Commit` concurrently), I'd rather hold off on any effort to restrict multiple `EndTxnRequest`. Of course, I would certainly feel differently if the raciness had any correctness implications.

---

*[pkg/internal/client/txn_test.go, line 517 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-Ke-N9hnBVKWBDFtgTaT:-KeCPbRlQoT0eDdiJB4t:b-ix40j6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5164a6b0b73ae0ecbc36f9727a9b134adb2ee993/pkg/internal/client/txn_test.go#L517)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`TransactionAbortedError` doesn't need to retain the key. `TransactionRetryError` *must* retain the key and *must not* send another `BeginTransaction` request. (That would require a pretty strange sequence of events; maybe it's a regression test for a bug we saw at some point?). So we want a transaction that writes to "first" on the first attempt (sending BeginTransaction, Put, and then getting a TransactionRetryError), then writes to "second" on the second iteration (sending Put with a Txn whose key points to "first") then commits. 
</blockquote></details>

Done. Thanks for the clarification.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 19 of 23 files at r10, 15 of 18 files at r11, 2 of 4 files at r12, 2 of 2 files at r13.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR. I'll wait for any remaining comments from Andrei.

How do we want to merge this change, given that it's pretty stability critical?

petermattis commented:
> How do we want to merge this change, given that it's pretty stability critical?

How risky do you feel the change is? Typical procedure recently is to either run a custom binary on one of the test clusters or to merge it on Thursday after the beta is released so we get maximal testing before the next beta.

bdarnell commented:
I'm not too concerned about this change since it just adds support for concurrency without actually introducing any concurrency, but we talked about it this morning and since we do have some other sensitive changes in the pipeline (lease rebalancing stuff is next in the queue), we'd like to hold off on merging this until after this week's beta (unless it's blocking other stuff)

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-:-KeZhwdy14liSLeO0U91:b78iw1l)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

But I had one question about the `TxnCoordSender` changes.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 878 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KeZsqVAOroojFJtak8-:-KeZsqVAOroojFJtak80:b-9k6udy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5b659a91340400cc7ceb96c5f35b86880e9c57e/pkg/kv/txn_coord_sender.go#L878)):*
> ```Go
> 			txnMeta.keys = keys
> 		} else if len(keys) > 0 {
> 			// For review: again, we need to take this check out. Is this the
> ```

I lost track of how the `Writing` field changed... Why exactly do we have to take these lines out, and the other panic below in this file?

---

*[pkg/kv/txn_coord_sender_test.go, line 963 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KeZsc0f3KOue5Lw-Vm4:-KeZsc0f3KOue5Lw-Vm5:b-8hij9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc4491733e21b6a1ed10e3098a1909ec72a7822/pkg/kv/txn_coord_sender_test.go#L963)):*
> ```Go
> 		ok      bool
> 	}{
> 		{roachpb.NewGet(roachpb.Key("a")), true, false},
> ```

nit: if you could take the opportunity to comment these bool args inline, that'd be fantastic

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-:-KeZtJ2nQkXxyD8wIDv9:bq18qj1)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
I think I feel ok about the change too since the added concurrency support should not affect the transaction lifecycle until that concurrency is actually exercised. If we release now, we should get a week of testing before the next beta, which would also be reassuring.

---

Review status: 0 of 38 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/kv/txn_coord_sender.go, line 878 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KeZsqVAOroojFJtak8-:-KerxoqxrukRtAhsYHu8:b-bltnn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5b659a91340400cc7ceb96c5f35b86880e9c57e/pkg/kv/txn_coord_sender.go#L878)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I lost track of how the `Writing` field changed... Why exactly do we have to take these lines out, and the other panic below in this file?
</blockquote></details>

`Writing` is set on a `Transaction` by a `*Replica` during the `BeginTransaction` handling. Because `BeginTxnRequests` can now be issues concurrently with other transactional write requests, those other requests will not have the flag set when they return their transaction protos.

An alternative might be to have `*Replicas` set this flag on all writing requests, but I'm not sure if this would be safe to do. For instance, I could imagine this approach causing problems when the `BeginTxnRequests` fails. I'm also not confident I understand all assumptions made by different parts of the system when this flag is set.

---

*[pkg/kv/txn_coord_sender_test.go, line 963 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-KeZsc0f3KOue5Lw-Vm4:-Kes0JNJyM-V3GYUmqrF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc4491733e21b6a1ed10e3098a1909ec72a7822/pkg/kv/txn_coord_sender_test.go#L963)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: if you could take the opportunity to comment these bool args inline, that'd be fantastic
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13799#-:-Kes0ZrkQ6G2FchuXJwS:b-908hje)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 39 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/internal/client/client_test.go, line 401 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/13799#-LCpWpPA22eum9_aY7Xe:-LCpWpPA22eum9_aY7Xf:bkqwl7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/53d22487f27bab5228b481cd0f5391eb9e923e9c/pkg/internal/client/client_test.go#L401)):*
> ```Go
> 						return
> 					}
> 					// Attempt to read outside of txn. We need to guarantee that the
> ```

so how exactly are we guaranteeing that the Begin has finished? I don't think we do... And I believe I do see txns aborted for this reason.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 39 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/internal/client/client_test.go, line 401 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/13799#-LCpWpPA22eum9_aY7Xe:-LD-Qiy2AnEGNbXd-p-P:b-95rmvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/53d22487f27bab5228b481cd0f5391eb9e923e9c/pkg/internal/client/client_test.go#L401)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

so how exactly are we guaranteeing that the Begin has finished? I don't think we do... And I believe I do see txns aborted for this reason.
</blockquote></details>

BeginTxn was executed as a part of txn.Put a few lines above. This comment only applies to this specific test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 39 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/internal/client/client_test.go, line 401 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/13799#-LCpWpPA22eum9_aY7Xe:-LD2EW8BAUqtSH4QAMd4:b3ft8dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/53d22487f27bab5228b481cd0f5391eb9e923e9c/pkg/internal/client/client_test.go#L401)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

BeginTxn was executed as a part of txn.Put a few lines above. This comment only applies to this specific test.
</blockquote></details>

Right, the call to `txn.Put` ensures that the txn record is written before we perform this read. The wording here is a little strange, feel free to re-word if you're touching this test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13799)*
<!-- Sent from Reviewable.io -->
